Access Control: Add service accounts

donomii opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR integrates code that will be the basis for further work on service accounts.  It is an interim release to prevent merge conflicts and to show the direction we are taking to implement service accounts.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana-enterprise/issues/1858

**Special notes for your reviewer**:

This is partial release to integrate the current work on service accounts with main.  The patch was growing too large to manage, so we are breaking the work into smaller releases.

This PR introduces the basic components needed to implement service accounts.  It changes the user and api_key tables, and the structures needed to deal with them.

It also modifies the "create apikey" api call, allowing for some basic testing of the entire patch.

The goal of this patch is not to add functionality, but to integrate our current work and prevent merge conflicts later, as well as showing the direction we are taking in implementing service accounts.

Commit:
WIP service accounts

hugohaggmark commented on public/app/features/api-keys/ApiKeysForm.tsx:
```diff
@@ -99,6 +119,17 @@ export const ApiKeysForm: FC<Props> = ({ show, onClose, onKeyAdded }) => {
+            <div className="gf-form">
+              <label>
+                New Service Account
+                <input
+                  name="createServiceAccount"
+                  type="checkbox"
+                  checked={createServiceAccount}
+                  onChange={onCreateServiceAccountChange}
+                />
```
nit: maybe change to more explicit boolean
```suggestion
                <input
                  name="createServiceAccount"
                  type="checkbox"
                  checked={Boolean(createServiceAccount)}
                  onChange={onCreateServiceAccountChange}
                />
```

hugohaggmark commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -76,6 +76,7 @@ export class ApiKeysPageUnconnected extends PureComponent<Props, State> {
   };
 
   onAddApiKey = (newApiKey: NewApiKey) => {
+    console.log(newApiKey);
```
Please remove

hugohaggmark reviewed (dismissed):
Left some comments about the frontend parts and also the UI needs more polish. I didn't see the `New Service Account` checkbox 
![image](https://user-images.githubusercontent.com/562238/132633433-070f71d8-c65b-4cde-a6cd-109ea1cacf4e.png)

Commit:
Update public/app/features/api-keys/ApiKeysForm.tsx

Co-authored-by: Hugo Häggmark <hugo.haggmark@grafana.com>

Commit:
Create a service account at the same time as the API key

kminehart commented:
@donomii I'd be happy to review this! I think a lot of what I'm looking at (in the backend) has a lot of "todo" markers. Since this is WIP, when you're ready for a review on the Go stuff, feel free to ping me!

donomii commented:
Thanks!

I made a mistake with Github and didn’t set the PR to draft, so it added a lot of reviewers to the PR.  Sorry about that!

It will be a couple of weeks before it’s ready for review, I’ll let everyone know when it is finished. 

> On 9. Sep 2021, at 22:19, Kevin Minehart ***@***.***> wrote:
> 
> 
> @donomii <https://github.com/donomii> I'd be happy to review this! I think a lot of what I'm looking at (in the backend) has a lot of "todo" markers. Since this is WIP, when you're ready for a review on the Go stuff, feel free to ping me!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/grafana/grafana/pull/38994#issuecomment-916408355>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABP7X2E25V4TRJOMOE7JRD3UBEJGZANCNFSM5DUZN3LQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

Commit:
Use service account credentials when accessing API with APIkey

Commit:
Throw better error

Commit:
Use Boolean for "create service account button"

Commit:
Add GetRole to service, merge RoleDTO and Role structs

This patch merges the identical OSS and Enterprise data structures, which improves the code for two reasons:

1.  Makes switching between OSS and Enterprise easier
2.  Reduces the chance of incompatibilities developing between the same functions in OSS and Enterprise

Commit:
Start work cloning permissions onto service account

Commit:
Add extra fields to OSS types to support enterprise

Commit:
If API key is not linked to a service account, continue login as usual

Commit:
Fallback to old auth if no service account linked to key

Commit:
Commented

Commit:
Add CloneUserToServiceAccount

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Update mock.go

Commit:
Put graphical bits behind a feature toggle

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Start adding LinkAPIKeyToServiceAccount

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Update pkg/models/user.go

Co-authored-by: Eric Leijonmarck <eric.leijonmarck@gmail.com>

Commit:
Update pkg/api/apikey.go

Co-authored-by: Eric Leijonmarck <eric.leijonmarck@gmail.com>

Commit:
Update pkg/api/apikey.go

Co-authored-by: Eric Leijonmarck <eric.leijonmarck@gmail.com>

Commit:
Finish LinkAPIKeyToServiceAccount

Commit:
Update comment

eleijonmarck reviewed (commented):
just one missing err I think

Commit:
Handle api key link error

eleijonmarck reviewed (dismissed):
![accounts](https://user-images.githubusercontent.com/7727602/136579896-12073ed1-198d-4ac3-8c9e-2a8e78a9d032.gif)

sakjur reviewed (dismissed):
I _think_ this breaks API keys from the perspective of the OSS codebase. Let's hold off on merging until that's cleared out.

I'm also a bit concerned about introducing panics, we have error handling in place in each of the places where panics are being introduced and I don't think any of them should be considered runtime errors, I think we should avoid panicking because we _can_ effectively recover from these errors. (see [effective go on panics](https://golang.org/doc/effective_go#panic))

Commit:
Update pkg/services/sqlstore/apikey.go

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

IevaVasiljeva reviewed (commented):
Some questions and suggestions in comments. Feel free to ignore the nits if they don't align with your personal preferences 🙂 

I'd also love to test the changes out - what's the best way to do it? Should I call the endpoints directly or is there a nice way to test them through the UI?

Commit:
Feature toggle

Commit:
Update pkg/services/accesscontrol/accesscontrol.go

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

Commit:
Not needed (yet)

Commit:
Better error messages for OSS accesscontrol

donomii commented:
> Some questions and suggestions in comments. Feel free to ignore the nits if they don't align with your personal preferences 🙂
> 
> I'd also love to test the changes out - what's the best way to do it? Should I call the endpoints directly or is there a nice way to test them through the UI?

I removed the UI for now, but I can add that back into another branch if you are interested.  For now you can call the API and set the CreateNewServiceAccount flag to true.  You can then look at the database table to see the new service account in the apikey table

donomii commented:
pkg/models/user.go
	Result           User
	IsServiceAccount bool
 
Member
@IevaVasiljeva IevaVasiljeva 23 hours ago
nit: is Result User separated from the rest of the fields because it is the result of the command? If so, should IsServiceAccount bool be defined together with the rest of the attributes higher up?

This is a left over habit I have from another programming language, where it is very important to always add new fields at the end of the structure.  But because we have a monolithic program, we can add them anywhere.

Commit:
Set an invalid user id as default

Commit:
ServiceAccountId should be string

Commit:
Re-arrange field names

Commit:
ServiceAccountId is integer

Commit:
Update ossaccesscontrol.go

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Linter

Commit:
Remove fronend edits

Commit:
Remove console log

Commit:
Update ApiKeysForm.tsx

marefr reviewed (commented):
Added some comments/questions. In general looks good, but feels like you've might added to much code in certain places that doesn't do anything and could affect existing functionality - please make sure to guard everywhere with the feature toggle if you want to keep things as is. 

Have #16424 been considered in relation to this?

Commit:
Remove unnecessary db query, move to right place

gamab reviewed (commented):
I just started looking at your work, so this is unrelated with the core work but I found some mismatches in the mock :slightly_smiling_face:

Commit:
Update pkg/services/accesscontrol/mock/mock.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

Commit:
Update pkg/services/accesscontrol/mock/mock.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

Commit:
Update pkg/services/accesscontrol/mock/mock.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

Commit:
Better error messages

Commit:
Better and correcter error messages

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

Commit:
Remove is_service_account column, for now

Commit:
Merge branch 'main' into WIP-jeremy/Service-accounts-and-apikeys

eleijonmarck reviewed (approved):
LGTM! looks amazing
